### PR TITLE
Split type-deduction lattice/constraints analysis from type analysis.

### PR DESCRIPTION
**Pull request by azreika**
Reduces the bloat in both significantly (`ast/analysis/TypeConstraints.h` created, taking in parts of `ast/analysis/Type.h`), will be much easier to maintain and work with. Resolves issue #1733. 

Should be merged AFTER PR #1725 (as it's a branch off of it).

### CODE REVIEW

File: src/ast/analysis/TypeConstraints.cpp [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+
+/**
+ * Constraint on record type and its elements.
+ */
+static TypeConstraint isSubtypeOfComponent(
+        const TypeVar& elementVariable, const TypeVar& recordVariable, size_t index) {
+    struct C : public Constraint<TypeVar> {
+        TypeVar elementVariable;
+        TypeVar recordVariable;
+        unsigned index;
```
Comment (mmcgr) [inline src/ast/analysis/TypeConstraints.cpp:320]: This is a little bizarre - index is std::size_t, int, and unsigned. Standardise on ... size_t.

File: src/ast/analysis/TypeConstraints.cpp [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+        // If so then add overloads as alternatives
+        if (!typeAnalysis.hasValidTypeInfo(intrFun))
+            addConstraint(satisfiesOverload(typeEnv, functorBuiltIn(intrFun->getBaseFunctionOp()), functorVar,
+                    argVars, isInfixFunctorOp(intrFun->getBaseFunctionOp())));
+
+        // In polymorphic case
+        // We only require arguments to share a base type with a return type.
+        // (instead of, for example, requiring them to be of the same type)
+        // This approach is related to old type semantics
+        // See #1296 and tests/semantic/type_system4
```
Comment (mmcgr) [inline src/ast/analysis/TypeConstraints.cpp:501]: This can probably be removed.
```suggestion
```

**Review by mmcgr — CHANGES_REQUESTED** [review]
Most of the comments are actually about moved code, rather than new code, so you may not want to address those here.

**Comment by azreika** [comment]
I'll address these code-style changes in the next type-analysis PR (they are all in code that has been moved).

**Review by mmcgr — APPROVED** [review]

**Comment by azreika** [comment]
In the clean-up PR, printing should be cleaned up for the type-related analyses too (i.e. the main type analysis + polymorphic + functor analyses).

**Pull request merged**